### PR TITLE
Copy from device after every trace execution in tt-transformers prefill

### PR DESCRIPTION
**Pull request by rdraskicTT**
We have an issue where the model fails determinism test even though seed is fixed. The culprit was logits corruption during prefill. Logits tensor was sometimes corrupted by the execution of trace by the following user, so the resolution was to copy the tensor to host before executing the next trace.

https://github.com/tenstorrent/tt-metal/issues/34346

- [x] [vllm nightly llama](https://github.com/tenstorrent/tt-metal/actions/runs/20260232100)
- [x] [vllm nightly qwen](https://github.com/tenstorrent/tt-metal/actions/runs/20297140813)
- [x] [llama 8B model ci, seed test passing](https://github.com/tenstorrent/tt-shield/actions/runs/20260818556/job/58174961837#step:9:3370) 
- [x] [t3k llama demo tests](https://github.com/tenstorrent/tt-metal/actions/runs/20273268045)

### CODE REVIEW

File: models/tt_transformers/tt/generator.py [hunk 1]
```diff
@@ -360,28 +358,21 @@ def prefill_forward_text(
...
-                output_logits[idx] = self.model[model_id].process_output_prefill(
-                    logits, last_token_idx=(last_token_idx % 32)
-                )
-                del logits
+            # We have to dispatch copy to host to avoid corruption by the next user's prefill
+            out_list.append(logits.cpu(blocking=False))
 
         # Process the logits after all the prefill are done in data parallel mode
-        if self.data_parallel > 1:
-            for idx, out in enumerate(out_list):
```
Comment (skhorasganiTT) [inline models/tt_transformers/tt/generator.py:375]: This if-else was added by @djordje-tt because in the DP=1 case there was an out of DRAM issue that was resolved by deleting the logits. We don't want to reintroduce that issue so maybe the DP=1 case should be left untouched.

File: models/tt_transformers/tt/model.py [hunk 2]
```diff
@@ -353,10 +353,10 @@ def concat_host_output(self, tt_out, is_log_probs=False):
 
     def process_output_prefill(self, tt_out, last_token_idx):
         """
-        Input is ttnn device tensor of logits. Output is torch logits tensor.
+        Input is ttnn host tensor of logits. Output is torch logits tensor.
         NOTE: In this model, prefill always uses get_last_token
         """
-        return self.concat_host_output(tt_out.cpu())[0, 0, last_token_idx, : self.vocab_size]
+        return self.concat_host_output(tt_out)[0, 0, last_token_idx, : self.vocab_size]
```
Comment (skhorasganiTT) [inline models/tt_transformers/tt/model.py:360]: Perhaps should assert here that tensor is on host

File: models/tt_transformers/tt/generator.py (hunk 1 above)
Comment (rdraskicTT) [reply models/tt_transformers/tt/generator.py:375]: That's a good point, I'm aware of that. I don't have a good answer, I prefer the path I took in this PR. The problem is that DP=1/TP=1 is broken again due to additional memory being used for trace capture. I plan to have a followup PR where we completely trace prefill (including `process_logits_after_prefill_trace`), in that case there will be no problem with extra logits tensors since they are being reused on device, although now we have less space due to trace being stored. Basically, given that we probably always want to have active trace and that we will also have batched-prefill support soon, this special case no longer makes sense since it won't have any of the latest perf improvements. I think that longterm, we either say we don't support it on n150, or reduce context length even further.

**Review by bgoelTT — COMMENTED** [review]
@rdraskicTT does this also address Llama 70B's failure to pass the non-uniform seed test?

**Comment by rdraskicTT** [comment]
> @rdraskicTT does this also address Llama 70B's failure to pass the non-uniform seed test?

@bgoelTT Haven't looked into 70B's code, this fixes only TTT models, but maybe 70B has a similar issue since prefill trace implementation was heavily inspired by 70B code.

File: models/tt_transformers/tt/model.py (hunk 2 above)
Comment (rdraskicTT) [reply models/tt_transformers/tt/model.py:360]: Good point, wasn't aware that this can easily be checked! Added assert

**Review by ppetrovicTT — APPROVED** [review]

**Comment by bgoelTT** [comment]
> > @rdraskicTT does this also address Llama 70B's failure to pass the non-uniform seed test?
> 
> @bgoelTT Haven't looked into 70B's code, this fixes only TTT models, but maybe 70B has a similar issue since prefill trace implementation was heavily inspired by 70B code.

@rdraskicTT yes this will definitely need to be fixed in llama 70B. It's currently failing the same non-uniform sampling test

**Review by skhorasganiTT — APPROVED** [review]

**Comment by yieldthought** [comment]
Presumably this hurts performance. How much? Is there a reason not to try to double buffer the logits, so we are reading from buffer1 whilst the next pass is writing to buffer2 and then we flip them?

**Comment by rdraskicTT** [comment]
> Presumably this hurts performance. How much? Is there a reason not to try to double buffer the logits, so we are reading from buffer1 whilst the next pass is writing to buffer2 and then we flip them?

@yieldthought I didn't notice any perf hit, I've compared with old runs. I suspect this is because I'm effectively only reordering ops. I thought about double buffering but I think that it's an overkill in this situation since model pass dominates, shaving off a millisecond is not worth the trouble at the moment imo. Also, the main goal of this PR is to fix the bug where logits get corrupted if they stay on device.

**Review by yieldthought — APPROVED** [review]
